src/ray/raylet/node_manager.cc

```diff
@@ -414,43 +404,56 @@ void NodeManager::Heartbeat() {
           ResourceSet(local_resources.GetTotalResources()));
     }
 
+    if (!last_heartbeat_resources_.GetAvailableResources().IsEqual(
+            local_resources.GetAvailableResources())) {
+      if (local_resources.GetAvailableResources().IsEmpty()) {
+        (*heartbeat_data->mutable_resources_available())[kResourcesTurningEmpty] = 0.0;
+      } else {
+        for (const auto &resource_pair :
+             local_resources.GetAvailableResources().GetResourceMap()) {
+          (*heartbeat_data->mutable_resources_available())[resource_pair.first] =
+              resource_pair.second;
+        }
+      }
+      last_heartbeat_resources_.SetAvailableResources(
+          ResourceSet(local_resources.GetAvailableResources()));
+    }
+
     local_resources.SetLoadResources(local_queues_.GetTotalResourceLoad());
     if (!last_heartbeat_resources_.GetLoadResources().IsEqual(
```

Comment:
This code is too nested. Why don't we combine two predicates?

```
if (! last_heartbeat_resources_. GetLoadResources(). IsEqual. ... && local_resources.GetLoadResources().IsEmpty()) {
} else {
}
```

Reply:
Same for everywhere that has nested 

```
if (A) {
  if (B) {
  
  } else {}
}
```

=>

```
if (A && B) {

} else {}
```

Reply:
No. It's not such a simple condition combination...
Conditions here are:
a) if A && B, do 1
b) if A && !B, do 2
c) if A, no matter what B is, do 3
d) if !A, no matter what B is, do nothing.
If we change the logic to
if (A && B) { do 1 } else { do 2}, we will miss c) and d)

Reply:
Ah, I see. That's right.